src/providers/trusted_service/context/key_management.rs

```diff
@@ -0,0 +1,146 @@
+                lifetime: KeyLifetime::Persistent as u32,
+                id,
+                policy: Some(KeyPolicy {
+                    usage: key_attrs.policy.usage_flags.try_into()?,
+                    alg: key_attrs.policy.permitted_algorithms.try_into()?,
+                }),
+            }),
+        });
+        let mut proto_resp = Box::new(GenerateKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::GenerateKey,
+            self.rpc_caller,
+        )?;
+
+        Ok(0)
+    }
+
+    pub fn destroy_key(&self, id: u32) -> parsec_interface::requests::Result<()> {
+        if !self.check_key_exists(id)? {
+            return Err(ResponseStatus::PsaErrorDoesNotExist);
+        }
+        let proto_req = Box::new(OpenKeyIn { id });
+        let mut proto_resp = Box::new(OpenKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::OpenKey,
+            self.rpc_caller,
+        )?;
+        let handle = proto_resp.handle;
+
+        let proto_req = Box::new(DestroyKeyIn { handle });
+        let mut proto_resp = Box::new(DestroyKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::DestroyKey,
+            self.rpc_caller,
+        )?;
+        Ok(())
+    }
+
+    pub fn check_key_exists(&self, id: u32) -> Result<bool, Error> {
+        let proto_req = Box::new(OpenKeyIn { id });
+        let mut proto_resp = Box::new(OpenKeyOut::default());
+        match Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::OpenKey,
+            self.rpc_caller,
+        ) {
+            Ok(_) => {
+                let handle = proto_resp.handle;
+                let proto_req = Box::new(CloseKeyIn { handle });
+                Context::send_request(
+                    &proto_req,
+                    &mut proto_resp,
+                    Opcode::OpenKey,
+                    self.rpc_caller,
+                )?;
+                Ok(true)
+            }
+            Err(e) => {
+                if e == Error::DoesNotExist {
+                    Ok(false)
+                } else {
+                    Err(e)
+                }
+            }
+        }
+    }
+
+    fn send_request<Req: Message, Resp: Message>(
+        req: &Box<Req>,
+        resp: &mut Box<Resp>,
+        opcode: Opcode,
+        rpc_cl: *mut rpc_caller,
+    ) -> Result<(), Error> {
+        info!("Beginning call to Trusted Service");
+        let mut buf_out = null_mut();
+
+        let call_handle = unsafe { rpc_caller_begin(rpc_cl, &mut buf_out, req.encoded_len()) };
+        if call_handle == null_mut() {
+            error!("Call handle was null");
+            return Err(Error::CommunicationFailure);
+        }
+        let mut buf_out = unsafe { Vec::from_raw_parts(buf_out, 0, req.encoded_len()) };
```

Comment:
Shouldn't length be equal to capacity?

Reply:
I'll try that again, but the weird thing is that when I tried, it was putting `req.encoded_len()` zeroes at the end, so the final length of the vector was actually twice what I wanted. 

It also doesn't help that the documentation only says about the 2nd field: `length needs to be less than or equal to capacity.`

Reply:
Resolving this as moving to `slice` removed that issue